### DISCUSSION TITLE
Option badwords not working with more than one word with pam conf

### DISCUSSION THREAD

**Discussion by prisme60** [libpwquality/libpwquality#90] [Q&A] [answered]
# Problem
I try to configure pwquality to forbid some words, but only the first word is working.

# Configuration
In my /etc/pam.d/common-password file:
```
password    requisite           /usr/lib/security/pam_pwquality.so \
                                   retry=3 minlen=14 dcredit=-1 ucredit=-1 ocredit=-1 lcredit=-1 enforce_for_root \
                                   badwords=microsoft google intel
```

# Tests with `passwd`
I try to use the following password : "Microsoft!1234567890"
It is refused because of the bad word "microsoft" (BAD PASSWORD: The password contains forbidden words in some form). 👍 

But if I use "Google!1234567890", the password is accepted. So google is not considered as a bad word. 👎 

In the log, I have the following errors in the log :
```
Sep 11 08:59:14 qemuarm passwd[247]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; google
Sep 11 08:59:14 qemuarm passwd[247]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; intel
```
It means that the space is used a separator for the arguments of pwquality, but when the library treats the option badwords, the value can't contains space.

# Documentation
In the documentation, it is written: 
```
badwords

Space separated list of words that must not be contained in the password. These
are additional words to the cracklib dictionary check. This setting can be
also used by applications to emulate the gecos check for user accounts that are
not created yet.
```

**But there is no example on the badwords option in the documentation.**

# Other tries but without success
## Parameter regroupment
I try to regroup my parameter with:
- simple quote
```'badwords=microsoft google intel'```

```log
Sep 11 09:21:34 qemuarm passwd[271]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; 'badwords=microsoft
Sep 11 09:21:34 qemuarm passwd[271]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; google
Sep 11 09:21:34 qemuarm passwd[271]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; intel'
```

or
```badwords='microsoft google intel'```
```log
Sep 11 09:22:42 qemuarm passwd[268]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; google
Sep 11 09:22:42 qemuarm passwd[268]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; intel'
```

- double quote
```"badwords=microsoft google intel"```

```log
Sep 11 09:25:37 qemuarm passwd[271]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; "badwords=microsoft
Sep 11 09:25:37 qemuarm passwd[271]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; google
Sep 11 09:25:37 qemuarm passwd[271]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; intel"
```

or
```badwords="microsoft google intel"```

```log
Sep 11 09:23:43 qemuarm passwd[268]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; google
Sep 11 09:23:43 qemuarm passwd[268]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; intel"
```

- escape space character
```badwords=microsoft\ google\ intel```

```log
Sep 11 09:28:46 qemuarm passwd[277]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; google\
Sep 11 09:28:46 qemuarm passwd[277]: pam_pwquality(passwd:chauthtok): pam_parse: unknown or broken option; intel
```

## Another separator (comma)
```badwords=microsoft,google,intel```

In this case, the word is very long `microsoft,google,intel`, because comma is not considered as a separator.

But I read the code, and we can see that a space is required.
https://github.com/libpwquality/libpwquality/blob/db3d72ecf0d95c34bd56f50574efbc6fe66cd176/src/check.c#L478

# Conclusion
In current state, the `badwords` option of libpwquality can't be used with more than one word with pam/passwd. I suggest to add comma as an additional separator.

**Comment by t8m** [comment, marked as answer]
You have to use the pam configuration escaping here.

This should work:

```
password    requisite           /usr/lib/security/pam_pwquality.so \
                                   retry=3 minlen=14 dcredit=-1 ucredit=-1 ocredit=-1 lcredit=-1 enforce_for_root \
                                   [badwords=microsoft google intel]
```

The `[]` characters are basically like double quotes in shell.

**Reply by prisme60** [reply]
Thank you for your really fast and accurate answer.

I just tested it, and it works with surrounding my argument with square brackets! 👍 

I suggest you to add an example in the pam_pwquality.8.pod file.
